compiler/luci/service/src/Nodes/CircleFullyConnected.cpp

```diff
@@ -36,4 +40,65 @@ luci::CircleNode *CloneNodeLet<CN::DEF>::visit(const luci::CircleFullyConnected
   return cloned;
 }
 
+namespace sinf
+{
+
+loco::TensorShape Algorithm::visit(const luci::CircleFullyConnected *node)
+{
+  auto input_shape = circle_shape(loco::must_cast<CircleNode *>(node->input()));
+  auto weights_shape = circle_shape(loco::must_cast<CircleNode *>(node->weights()));
+
+  loco::TensorShape out_shape;
+
+  // NOTE Some recipes in some repositories are using rank 4 input for FullyConnected.
+  //      Until they are all fixed, disable following assert.
+  // TODO Enable following assert after related fixes are applied
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L194
+  // LUCI_ASSERT(input_shape.rank() == 2 || input_shape.rank() == 3,
+  //             "Input rank of FullyConnected should be 2 or 3");
+
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L225
+  LUCI_ASSERT(weights_shape.rank() == 2, "Weights of FullyConnected should be 2");
+
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L353-L367
+  if (node->keep_num_dims())
+  {
+    out_shape.rank(input_shape.rank());
+    for (uint32_t i = 0; i < input_shape.rank(); ++i)
+      out_shape.dim(i) = input_shape.dim(i);
+    out_shape.dim(out_shape.rank() - 1) = weights_shape.dim(0);
+  }
+  else
+  {
+    bool is_dynamic_shape = false;
+
+    for (uint32_t i = 0; i < input_shape.rank() - 1; i++)
+    {
+      if (not input_shape.dim(i).known())
+      {
+        is_dynamic_shape = true;
+        break;
+      }
+    }
+
+    uint32_t batch_size = 1;
+
+    for (uint32_t i = 0; i < input_shape.rank() - 1; i++)
+    {
+      batch_size *= input_shape.dim(i).value();
+    }
+
+    out_shape.rank(2);
+    if (is_dynamic_shape)
+      out_shape.dim(0).unset();
+    else
+      out_shape.dim(0) = batch_size;
+    out_shape.dim(1) = weights_shape.dim(0);
+  }
```

Comment:
( I don't want to force it at all. So plz remark this is just a suggestion. ) 

How about writing codes in this way? 

To write under codes, we have to define `remove_last_dim()` and `is_unknown()` functions in the anonynmous namespace.

```suggestion
    TensorShape input_without_last = remove_last_dim(input_shape)
    
    out_shape.rank(2);
    if (is_unknown(input_without_last)) 
      out_shape.dim(0).unset();
    else
      out_shape.dim(0) = input_without_last.element_count()
    
    out_shape.dim(1) = weights_shape.dim(0);
```
```